### DISCUSSION TITLE
Can I use loader data in a route's links function?

### DISCUSSION THREAD

**Discussion by kishba** [remix-run/remix#1359] [Q&A] [answered]
I was trying to make a dynamic favicon on one of my nested routes. (More specifically, my idea was to display a different color favicon in staging environments.)

Is it possible to access loader data in a meta function I define on a route? My first few attempts tried to use loader data to get the request URL so I could inspect the domain name and it sorta worked... until React hydrated. :)

EDIT: When I wrote "meta function" I should have written "links function" -- sorry about this mixup!

**Comment by sergiodxa** [comment, marked as answer]
The MetaFunction receives a single object argument, that object has a data property with the data returned by the loader.

You can’t access the request but if you know you need the url or some search param from there you can return it from the loader and access it inside the data property.

Note that MetaFunction doesn’t receive a generic so data is typed as any, you should cast the type or validate it somehow to narrow the type.

**Reply by kishba** [reply]
I appreciate your reply @sergiodxa!

In my question I mentioned MetaFunction, but I think I should have said the `links` function. That's where I'm currently creating my favicon entry. Can loader data be used inside a `links` function? When I try, I get the following error:

```
[1] TypeError: Cannot destructure property 'data' of 'undefined' as it is undefined.
```

Here are my functions if that helps...

```ts
function favicon(environment: string) {
  switch (environment) {
    case "development": {
      return "/favicons/dev-favicon.ico";
    }
    default: {
      return "/favicons/favicon.ico";
    }
  }
}

export function links({ data }) {
  console.log(data);
  return [
    { rel: "stylesheet", href: styles },
    {
      rel: "icon",
      href: favicon("production") // TODO: how can I use loader data in the links function?
    }
  ];
}
```

**Reply by sergiodxa** [reply]
You can’t on the LinksFunction because Remix needs it to be stable so it can be prefetched.

I added a feature on Remix Utils for that https://github.com/sergiodxa/remix-utils#dynamiclinks, the links there are not going to be prefetched but you can use dynamic links.

**Reply by kishba** [reply]
Oh, cool, I'll check that out!

**Reply by kishba** [reply]
@sergiodxa before I tried your DynamicLinks component, I added the environment variable I care about to `window` and then checked to see if either `window` or `process.env` was available my `links` function. Does this introduce instability in my links function like what you referred to?

```js
export function links() {
  let deploymentEnvironment;
  // server
  if (typeof process !== "undefined") {
    deploymentEnvironment = process.env.SC_DEPLOY_ENV;
  }
  // client
  if (typeof window !== "undefined") {
    deploymentEnvironment = window.ENV.SC_DEPLOY_ENV;
  }
  return [
    { rel: "stylesheet", href: styles },
    {
      rel: "icon",
      href: favicon(deploymentEnvironment)
    }
  ];
}
```

**Reply by sergiodxa** [reply]
I wouldn’t recommend you to do this, it looks like something easy to break.

Since the favicon doesn’t usually change between routes, you could return the env variable from the loader of your root route, then where you render the head tag render the link to the favicon instead of using the LinksFunction

**Reply by GuyoST** [reply]
> You can’t on the LinksFunction because Remix needs it to be stable so it can be prefetched.
> 
> I added a feature on Remix Utils for that https://github.com/sergiodxa/remix-utils#dynamiclinks, the links there are not going to be prefetched but you can use dynamic links.

Hello, I tried it but I got some trouble to make it work, I posted an example there : 
https://github.com/sergiodxa/remix-utils/discussions/98

Can you help me ?
